Convert Scratch lookup to binary search

amanasifkhalid opened the pull request:
Preliminary work for #1903: Before making any design changes, we should try using binary search over the Scratch table, and profile the performance gain. I likely won't have time to fully pursue this task, but we can at least get rid of the linear search for now.

Commit:
Convert Scratch lookup to binary search

cshung commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -5756,6 +5756,84 @@ int NativeUnwindInfoLookupTable::LookupUnwindInfoForMethod(DWORD RelativePc,
+    // Casting the lookup table's size to an int is safe:
+    // We index the RUNTIME_FUNCTION table with ints, and the lookup table
+    // contains a subset of the indices in the RUNTIME_FUNCTION table.
+    // Thus, |lookup table| <= |RUNTIME_FUNCTION table|.
```
Why not just assert that fact instead?

amanasifkhalid replied:
I can add an assert for this -- we'll just pass `pInfo` instead of `m_nScratch` and `m_pScratch` separately so we can access the size of the runtime function table.

cshung commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -5756,6 +5756,84 @@ int NativeUnwindInfoLookupTable::LookupUnwindInfoForMethod(DWORD RelativePc,
+    // contains a subset of the indices in the RUNTIME_FUNCTION table.
+    // Thus, |lookup table| <= |RUNTIME_FUNCTION table|.
+
+    int high = ((nLookupTable - 1) / 2);
```
`nLookupTable` is always even, we can assert that fact. Should we really subtract 1 here?

amanasifkhalid replied:
By copying the binary search implementation in `NativeUnwindInfoLookupTable::LookupUnwindInfoForMethod`, I made `low` and `high` inclusive bounds, so I think subtracting 1 is required here. Without this, it is possible for us to index out of bounds in the `for` loop below (since we iterate while `i <= high`).

cshung replied:
Got it, would be more clear to write `nLookupTable/2 - 1` so that I know you want an inclusive upper bound, but that's okay.

cshung commented on src/coreclr/vm/codeman.h:
```diff
@@ -1499,6 +1499,14 @@ class NativeUnwindInfoLookupTable
+class HotColdMappingLookupTable
+{
+public:
+    static int LookupMappingForMethod(PTR_ULONG pLookupTable,
```
Can we have a comment saying what this function is supposed to return?

amanasifkhalid replied:
Sure thing.

amanasifkhalid commented:
@cshung I pushed changes addressing your feedback. For asserting the lookup table's size is <= the runtime function table's size, I changed the method's parameter list to just pass `pInfo`, but this required making `HotColdMappingLookupTable` a friend of `ReadyToRunInfo` -- not sure if this is ideal or not...

cshung commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -5756,6 +5756,84 @@ int NativeUnwindInfoLookupTable::LookupUnwindInfoForMethod(DWORD RelativePc,
+    // We index the RUNTIME_FUNCTION table with ints, and the lookup table
+    // contains a subset of the indices in the RUNTIME_FUNCTION table.
+    // Thus, |lookup table| <= |RUNTIME_FUNCTION table|.
+    _ASSERTE(pInfo->m_nScratch <= pInfo->m_nRuntimeFunctions);
```
While we are at it, we might as well `_ASSERTE(pInfo->m_nScratch % 2 == 0)`

amanasifkhalid replied:
Gotcha, just pushed a revision.
